### PR TITLE
Copy nn.{softmax,log_softmax} to scipy.special

### LINKED ISSUE #20700: Move jax.nn.log_softmax to jax.scipy.special?
It's in SciPy now, so it seems like the logical place for it?

### PR DESCRIPTION
**Pull request by NeilGirdhar**
Fixes #20700

### CODE REVIEW

**Comment by NeilGirdhar** [comment]
@jakevdp

File: jax/scipy/special.py [hunk 1]
```diff
@@ -15,6 +15,11 @@
 # Note: import <name> as <name> is required for names to be exported.
 # See PEP 484 & https://github.com/google/jax/issues/7570
 
+# These functions are exposed in both jax.nn (for backwards compatibility), and
```
Comment (hawkinsp) [inline jax/scipy/special.py:18]: Note: I would *not* consider exposing these in `jax.nn` a deprecated behavior. I encourage users to prefer the `jax.nn` copy to avoid the `scipy` import required by `jax.scipy`.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good! We'll also need a test (just a smoketest is fine, with a pointer to the full tests in `nn_test.py`) and also we need to add these new functions to the docs in this file: https://github.com/google/jax/blob/main/docs/jax.scipy.rst

**Review by jakevdp — APPROVED** [review]
Looks good!

File: jax/scipy/special.py [hunk 2]
```diff
@@ -15,6 +15,11 @@
...
+# jax.scipy.special (to support the Array API specification).
```
Comment (jakevdp) [inline jax/scipy/special.py:19]: Is this comment correct? I don't think the Array API says anything about scipy APIs.

Comment (NeilGirdhar) [reply jax/scipy/special.py:19]: You're right.  However, maybe I'm jumping the gun, but it [probably will soon enough](https://github.com/data-apis/array-api/issues/725).  They are almost surely going to add a "special functions" extension, and these functions are in the draft there.

Also, from the linked table, they appear to be in `scipy.special`, `torch.special`, and `cupyx.scipy.special`, so there's a consistency argument for having them there?

Comment (hawkinsp) [reply jax/scipy/special.py:19]: As I read it, it proposes a `special` namespace, not a `scipy.special` namespace.

**Comment by jakevdp** [comment]
It looks like these functions can't be used directly, because `jax.nn.softmax` defaults to `axis=-1`, while `scipy.special.softmax` defaults to `axis=None`.

We'll have to create wrappers for these in `jax/_src/scipy/special.py`.

File: jax/scipy/special.py (hunk 2 above)
Comment (NeilGirdhar) [reply jax/scipy/special.py:19]: @hawkinsp I'm assuming that if the `special` namespace is added, Jax will simply link it to `jax.scipy.special`?

**Comment by NeilGirdhar** [comment]
> It looks like these functions can't be imported directly, because `jax.nn.softmax` defaults to `axis=-1`, while `scipy.special.softmax` defaults to `axis=None`.
> 
> We'll have to create wrappers for these in `jax/_src/scipy/special.py`.

Ah, got it, so this is more tricky than I thought.

File: jax/scipy/special.py (hunk 2 above)
Comment (jakevdp) [reply jax/scipy/special.py:19]: It really depends what the final specification is.

But for now, adding these functions to `jax.scipy.special` is well in-scope for JAX (see https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html#axis-1-xla-alignment)

Comment (NeilGirdhar) [reply jax/scipy/special.py:19]: Okay, got it.  I was just trying to be proactive 😄   What I'll do then is try to produce shims with the appropriate axis behaviour, right?

Comment (jakevdp) [reply jax/scipy/special.py:19]: Yeah, that sounds great.

**Comment by francois-rozet** [comment]
Hello @NeilGirdhar, @jakevdp,

I have noticed a (very serious) bug with `jax.nn.softmax` which might be worse fixing within this PR.

```python
import jax
import jax.numpy as jnp

x = jax.random.normal(jax.random.key(0), (3, 1, 1))

def f(x):
    return jax.nn.softmax(x, axis=0)

print(f(x))
print(jax.jit(f)(x))
```

```
[[[0.75250584]]

 [[0.0755428 ]]

 [[0.17195134]]]
[[[1.]]

 [[1.]]

 [[1.]]]
```

Note that the shape of `x` seems to cause the bug within `jax.jit(f)`.

Using a "numerically safe" version of `softmax` based on `log_softmax` solves the issue.

```python
def f(x):
    return jnp.exp(jax.nn.log_softmax(x, axis=0))

print(f(x))
print(jax.jit(f)(x))
```


```
[[[0.75250584]]

 [[0.07554279]]

 [[0.17195135]]]
[[[0.75250584]]

 [[0.07554279]]

 [[0.17195135]]]
```

**Comment by NeilGirdhar** [comment]
@francois-rozet I think you should file a separate bug?

**Comment by francois-rozet** [comment]
@NeilGirdhar Ok I filled a bug report (#20856).

**Comment by jakevdp** [comment]
> We'll have to create wrappers for these in `jax/_src/scipy/special.py`.

Also, we might want to make the `jax.scipy.special` wrappers use the non-deprecated softmax version by default. The deprecated version has some poor autodiff behavior, but we haven't been able to turn it on by default in `jax.nn.softmax` because it changes the numerics and breaks some downstream projects that are sensitive to the change.

**Comment by jakevdp** [comment]
Are you still interested in contributing this?

**Comment by NeilGirdhar** [comment]
@jakevdp Yes, sorry for the delay.  I've been trying to find time to do it.

**Comment by NeilGirdhar** [comment]
@jakevdp  should be ready for your review.

File: jax/_src/scipy/special.py [hunk 3]
```diff
@@ -2582,3 +2584,78 @@ def hyp1f1(a: ArrayLike, b: ArrayLike, x: ArrayLike) -> Array:
...
+    \mathrm{log\_softmax}(x)_i = \log \left( \frac{\exp(x_i)}{\sum_j \exp(x_j)}
+    \right)
+
+  Args:
+    x : input array
+    axis: the axis or axes along which the :code:`log_softmax` should be
+      computed.
+
+  Returns:
+    An array.
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:2649]:
```suggestion
    An array of the same shape as ``x``
```

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/special.py [hunk 4]
```diff
@@ -2582,3 +2584,78 @@ def hyp1f1(a: ArrayLike, b: ArrayLike, x: ArrayLike) -> Array:
...
+  Note:
+    If any input values are ``+inf``, the result will be all ``NaN``: this
+    reflects the fact that ``inf / inf`` is not well-defined in the context of
+    floating-point math.
+
+  See also:
+    :func:`log_softmax`
+  """
+  if isinstance(axis, int) and axis < 0:
+    raise ValueError
```
[the remark this answers is not in the document]
Comment (NeilGirdhar) [reply jax/_src/scipy/special.py:2622]: Supporting -1 seems like it will be a giant bug magnet since [the Array API treats -1 as the last axis](https://data-apis.org/array-api/latest/API_specification/indexing.html#single-axis-indexing).    Why not just have users use `None`?

File: jax/_src/scipy/special.py [hunk 5]
```diff
@@ -2582,3 +2584,78 @@ def hyp1f1(a: ArrayLike, b: ArrayLike, x: ArrayLike) -> Array:
   lambda b_dot, primal_out, a, b, x: _hyp1f1_b_derivative(a, b, x) * b_dot,
   lambda x_dot, primal_out, a, b, x: _hyp1f1_x_derivative(a, b, x) * x_dot
 )
+
+
+# This interface is designed to be compatible with the future Array API
+# interface, which has not been specified yet as of June 2024.  We immitate its
+# sum function's use of keyword-only and positional-only parameters.
```
[the remark this answers is not in the document]
Comment (NeilGirdhar) [reply jax/_src/scipy/special.py:2591]: Okay, will remove.  Was just worried someone would change it at some point between now and when the Array API comes out :smile:

File: jax/_src/scipy/special.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/scipy/special.py:2624]: We're not implementing the array API here, we're implementing the `scipy.special` API.

[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/scipy/special.py:2621]: And my understanding is that `axis=None` is equivalent to `axis=tuple(range(x.ndim))`, so `axis=-1` is quite different than `axis=None`.

[the remark this answers is not in the document]
Comment (NeilGirdhar) [reply jax/_src/scipy/special.py:2621]: Okay, you're right.

**Comment by NeilGirdhar** [comment]
Is this ready to squash?

File: jax/_src/scipy/special.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/scipy/special.py]: I expect the array APi will likely be implemented in a different namespace, so even if this changes it won't be problematic.

File: tests/lax_scipy_special_functions_test.py [hunk 6]
```diff
@@ -261,5 +266,8 @@ def testBetaParameterDeprecation(self):
         lsp_special.beta(b=1, y=1)
 
 
+# softmax and log_softmax are tested in tests/nn_test.py.
```
Comment (jakevdp) [inline tests/lax_scipy_special_functions_test.py:269]: I think you can remove this comment. Most functions here are tested via `op_record`, so we don't need to call out that they're not tested via other means.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by NeilGirdhar** [comment]
Thanks for the speedy and patient review!
